geomstats/stiefel.py

```diff
@@ -0,0 +1,351 @@
+        block = gs.concatenate([AR, RZ], axis=1)
+        MNe = gs.expm(block)
+
+        return gs.matmul(
+            gs.concatenate([base_point, Q], axis=2), MNe[:, :, 0:p])
+
+    def log(self, point, base_point, max_iter=100, tol=1e-6):
+        """
+        Riemannian logarithm of a point wrt a base point.
+
+        Based on:
+        Zimmermann, Ralf
+        "A Matrix-Algebraic Algorithm for the Riemannian Logarithm
+        on the Stiefel Manifold under the Canonical Metric"
+        SIAM J. Matrix Anal. & Appl., 38(2), 322–342.
+        """
+        point = gs.to_ndarray(point, to_ndim=3)
+        n_points, _, _ = point.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_points == n_base_points
+                or n_points == 1
+                or n_base_points == 1)
+
+        M = gs.matmul(gs.transpose(base_point, (0, 2, 1)), point)
+
+        # QR of normal component of a point
+        K = point - gs.matmul(base_point, M)
+
+        Q = gs.zeros((K.shape))
+        N = gs.zeros((K.shape[0], K.shape[2], K.shape[2]))
+        for i, k in enumerate(K):
+            Q[i], N[i] = gs.linalg.qr(k)
+
+        # orthogonal completion
+        W = gs.concatenate([M, N], axis=1)
+
+        V = gs.zeros((
+            W.shape[0],
+            max(W.shape[1], W.shape[2]),
+            max(W.shape[1], W.shape[2])
+            ))
+
+        for i, w in enumerate(W):
+            V[i], _ = gs.linalg.qr(w, mode="complete")
+
+        # Procrustes preprocessing
+        [D, S, R] = gs.linalg.svd(V[:, p:2*p, p:2*p])
+
+        V[:, :, p:2*p] = gs.matmul(
+            V[:, :, p:2*p], gs.matmul(R, gs.transpose(D, axes=(0, 2, 1))))
+        V = gs.concatenate(
+            [gs.concatenate([M, N], axis=1), V[:, :, p:2*p]], axis=2)
+
+        for k in range(max_iter):
+
+            LV = gs.logm(V)
+
+            C = LV[:, p:2*p, p:2*p]
+            normC = gs.linalg.norm(C, ord=2, axis=(1, 2))
+
+            if normC < tol:
+                # print("Converged in {} iterations".format(k+1))
+                break
+
+            Phi = gs.expm(-C)
+            V[:, :, p:2*p] = gs.matmul(V[:, :, p:2*p], Phi)
+
+        XV = gs.matmul(base_point, LV[:, 0:p, 0:p])
+        QV = gs.matmul(Q, LV[:, p:2*p, 0:p])
+
+        return XV + QV
+
+    def retraction(self, tangent_vec, base_point):
+        """
+        Retraction map, based on QR-decomposion:
+        P_x(V) = qf(X + V)
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
+        n_tangent_vecs, _, _ = tangent_vec.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_tangent_vecs == n_base_points
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        # Q, R = gs.linalg.qr(base_point + tangent_vec)
+        # TODO: remove cycle, when qr will be vectorized
```

Comment:
`TODO(oleg)` makes it easier to grep what you need to work on (for example if we do a fix-it day or something). Thanks!